lib/routes/account.js

```diff
@@ -1045,50 +1128,84 @@ module.exports = function (
         }
       },
       handler: function (request, reply) {
-        log.begin('Account.RecoveryEmailVerify', request)
         var uid = request.payload.uid
         var code = Buffer(request.payload.code, 'hex')
         var service = request.payload.service || request.query.service
+
+        log.begin('Account.RecoveryEmailVerify', request)
         db.account(Buffer(uid, 'hex'))
           .then(
             function (account) {
-              // If the account is already verified, they may be e.g.
-              // clicking a stale link.  Silently succeed.
-              if (account.emailVerified) {
-                log.increment('account.already_verified')
-                return true
-              }
-              if (!butil.buffersAreEqual(code, account.emailCode)) {
-                throw error.invalidVerificationCode()
-              }
-              log.timing('account.verified', Date.now() - account.createdAt)
-              log.event('verified', request, {
-                email: account.email,
-                uid: account.uid,
-                locale: account.locale
-              })
-              log.increment('account.verified')
 
-              // send a push notification to all devices that the account changed
-              push.notifyUpdate(uid, 'accountVerify')
+              /**
+               * Logic for account and token verification
+               *
+               * 1) Attempt to use code as tokenVerificationId to verify session.
+               *
+               * 2) An error is thrown if tokenVerificationId does not exist (check to see if email
+               *    verification code) or the tokenVerificationId does not correlate to the
+               *    account uid (damaged linked/spoofed account)
+               *
+               * 3) Verify account email if not already verified.
+               */
+              return db.verifyToken(code, account)
+                .then(function () {
+                  log.event('tokenVerified', request, {
+                    email: account.email,
+                    uid: account.uid,
+                    locale: account.locale,
+                    tokenVerificationId: code
+                  })
+                })
+                .catch(function (err) {
```

Comment:
Yea, I think we should. While adding this in, it reminded me of an issue that @shane-tomlinson mentioned a while back that slipped out of my memory. Namely, that fxa-db-mysql returns 116 for [not found](https://github.com/mozilla/fxa-auth-db-mysql/blob/master/fxa-auth-db-server/lib/error.js#L31), but in fxa-auth-server, 116 is [endpoint not supported](https://github.com/mozilla/fxa-auth-server/blob/master/lib/error.js#L17).

Do errno in fxa-mysql and fxa-auth server have to be related or should I add a function to Errors to relate fxa-mysql and fxa-auth errno?